File: sdk/villow-sdk-0.1.0/package/src/testing.ts
import type { FastifyInstance } from 'fastify';

import { stableJsonBuffer, type Agent } from './agent.js';
import { signRequest } from './signing.js';

export interface HarnessOptions {
  renderMode?: 'rich' | 'simple';
  keyId?: string;
  publisherId?: string;
  agentId?: string;
  secret?: string;
}

export class MockPlatformHarness {
  private app: FastifyInstance;
  private counter = 0;
  private renderMode: 'rich' | 'simple';
  private keyId: string;
  private publisherId: string;
  private agentId: string;
  private secret: string;

  constructor(private agent: Agent, options: HarnessOptions = {}) {
    this.app = agent.fastifyApp();
    this.renderMode = options.renderMode ?? 'rich';
    this.keyId = options.keyId ?? agent.keyId;
    this.publisherId = options.publisherId ?? agent.publisherId;
    this.agentId = options.agentId ?? agent.agentId;
    this.secret = options.secret ?? agent.secret;
  }

  async runTask(
    taskTemplate: string,
    inputs: Record<string, unknown>,
    options: { answers?: Record<string, unknown>; delegatedDecisions?: Record<string, boolean> } = {},
  ): Promise<Record<string, unknown>> {
    const taskId = `task_${Date.now()}_${++this.counter}`;
    await this.post('/discover', {});
    let prepare = await this.post('/prepare', { task_id: taskId, task_template: taskTemplate, initial_inputs: inputs });
    const compositionResponse = prepare.composition_response as Record<string, unknown> | undefined;

    if (prepare.state === 'clarification_required') {
      const answers = options.answers ?? answersFromDecideForMe(prepare.questions as Array<Record<string, unknown>>);
      await this.post('/clarification_response', {
        task_id: taskId,
        clarification_id: prepare.clarification_id,
        phase: prepare.phase,
        answers,
        delegated_decisions: options.delegatedDecisions ?? {},
      });
      prepare = await this.post('/prepare', { task_id: taskId, task_template: taskTemplate, initial_inputs: inputs });
    }

    await this.post('/execute', {
      task_id: taskId,
      task_template: taskTemplate,
      inputs: prepare.normalized_inputs ?? inputs,
      tool_access_grants: [],
      callback_urls: {},
      sample_mode: false,
    });
    const result = await this.post('/result', { task_id: taskId });
    if (Array.isArray(result.artifacts) && result.artifacts.length > 0) result.state = 'completed';
    if (compositionResponse) {
      result.composition_response = compositionResponse;
      if (this.renderMode === 'simple') result.simple_fallback_slots = compositionResponse.fallback_slots;
    }
    return result;
  }

  async post(path: string, payload: Record<string, unknown>): Promise<Record<string, unknown>> {
    const body = stableJsonBuffer(payload);
    const idempotencyKey = `harness_${++this.counter}`;
    const headers = signRequest({
      method: 'POST',
      path,
      body,
      keyId: this.keyId,
      secret: this.secret,
      publisherId: this.publisherId,
      agentId: this.agentId,
      idempotencyKey,
    });
    headers['content-type'] = 'application/json';
    const response = await this.app.inject({ method: 'POST', url: path, payload: body, headers });
    if (response.statusCode < 200 || response.statusCode >= 300) {
      throw new Error(`harness_request_failed:${path}:${response.statusCode}:${response.body}`);
    }
    return response.json() as Record<string, unknown>;
  }

  async close(): Promise<void> {
    await this.app.close();
  }
}

function answersFromDecideForMe(questions: Array<Record<string, unknown>>): Record<string, unknown> {
  return Object.fromEntries(questions.map((question) => {
    const decision = question.decide_for_me as Record<string, unknown> | undefined;
    return [String(question.id), decision && Object.hasOwn(decision, 'value') ? decision.value : question.default_value];
  }));
}

File: sdk/villow-0.1.0/villow-0.1.0/src/villow/testing.py
from __future__ import annotations

import json
import time
from typing import Any

from fastapi.testclient import TestClient

from villow.signing import sign_request


class MockPlatformHarness:
    def __init__(
        self,
        agent,
        *,
        render_mode: str = "rich",
        key_id: str = "key_local",
        publisher_id: str = "publisher_local",
        agent_id: str = "agent_local",
        secret: str = "local-secret",
    ) -> None:
        self.agent = agent
        self.render_mode = render_mode
        self.key_id = key_id
        self.publisher_id = publisher_id
        self.agent_id = agent_id
        self.secret = secret
        self.client = TestClient(agent.asgi_app())
        self._counter = 0

    def run_task(
        self,
        task_template: str,
        inputs: dict[str, Any],
        *,
        answers: dict[str, Any] | None = None,
        delegated_decisions: dict[str, bool] | None = None,
    ) -> dict[str, Any]:
        task_id = f"task_{int(time.time() * 1000)}"
        self.post("/discover", {})
        prepare = self.post("/prepare", {"task_id": task_id, "task_template": task_template, "initial_inputs": inputs})
        composition_response = prepare.get("composition_response")

        if prepare.get("state") == "clarification_required":
            resolved_answers = answers or _answers_from_decide_for_me(prepare["questions"])
            self.post(
                "/clarification_response",
                {
                    "task_id": task_id,
                    "clarification_id": prepare["clarification_id"],
                    "phase": prepare["phase"],
                    "answers": resolved_answers,
                    "delegated_decisions": delegated_decisions or {},
                },
            )
            prepare = self.post("/prepare", {"task_id": task_id, "task_template": task_template, "initial_inputs": inputs})
            composition_response = prepare.get("composition_response") or composition_response

        self.post(
            "/execute",
            {
                "task_id": task_id,
                "task_template": task_template,
                "inputs": prepare.get("normalized_inputs", inputs),
                "tool_access_grants": [],
                "callback_urls": {},
                "sample_mode": False,
            },
        )
        result = self.post("/result", {"task_id": task_id})
        result["state"] = "completed" if result.get("artifacts") else result.get("state", "completed")
        if composition_response:
            result["composition_response"] = composition_response
            if self.render_mode == "simple":
                result["simple_fallback_slots"] = composition_response.get("fallback_slots", [])
        return result

    def post(self, path: str, payload: dict[str, Any]) -> dict[str, Any]:
        self._counter += 1
        body = json.dumps(payload, separators=(",", ":"), sort_keys=True).encode("utf-8")
        headers = sign_request(
            method="POST",
            path=path,
            body=body,
            key_id=self.key_id,
            secret=self.secret,
            publisher_id=self.publisher_id,
            agent_id=self.agent_id,
            timestamp=int(time.time()),
            nonce=f"harness_{self._counter}",
            idempotency_key=f"harness_{self._counter}",
        )
        headers["content-type"] = "application/json"
        response = self.client.post(path, content=body, headers=headers)
        response.raise_for_status()
        return response.json()


def _answers_from_decide_for_me(questions: list[dict[str, Any]]) -> dict[str, Any]:
    answers: dict[str, Any] = {}
    for question in questions:
        decision = question.get("decide_for_me") or {}
        if "value" in decision:
            answers[question["id"]] = decision["value"]
        elif question.get("default_value") is not None:
            answers[question["id"]] = question["default_value"]
        else:
            answers[question["id"]] = None
    return answers
